game/scripts/vscripts/components/duels/duels.lua

```diff
@@ -89,278 +44,106 @@ function Duels:StartDuel ()
-
-    goodGuy.duelNumber = 1
-    badGuy.duelNumber = 1
-
-    FindClearSpaceForUnit(goodHero, spawn1, true)
-    FindClearSpaceForUnit(badHero, spawn2, true)
-
-    Duels.zone1.addPlayer(goodGuy.id)
-    Duels.zone1.addPlayer(badGuy.id)
-
-    Duels:MoveCameraToPlayer(goodGuy.id, goodHero)
-    Duels:MoveCameraToPlayer(badGuy.id, badHero)
+  return {spawn1, spawn2}
+end
 
-    -- disable respawn
-    goodHero:SetRespawnsDisabled(true)
-    badHero:SetRespawnsDisabled(true)
+function Duels:CreateDuels()
+  local maxPlayers = 20
+  for _, team in self.teams do
+    if #team < maxPlayers then
+      maxPlayers = #team
+    end
   end
 
-  spawn1 = Entities:FindByName(nil, 'duel_2_spawn_1'):GetAbsOrigin()
-  spawn2 = Entities:FindByName(nil, 'duel_2_spawn_2'):GetAbsOrigin()
+  DebugPrint('Max players per team for this duel ' .. maxPlayers)
 
-  if spawnLocations then
-    local tmp = spawn1
-    spawn1 = spawn2
-    spawn2 = tmp
+  if maxPlayers < 1 then
+    DebugPrint('There aren\'t enough players to start the duel')
+    Notifications:TopToAll({text="There aren\'t enough players to start the duel", duration=2.0})
+    return
   end
 
-  for playerNumber = playerSplitOffset+1,maxPlayers do
-    DebugPrint('Checking player number ' .. playerNumber)
-    local goodGuy = Duels:GetUnassignedPlayer(goodPlayers, goodPlayerIndex)
-    local badGuy = Duels:GetUnassignedPlayer(badPlayers, badPlayerIndex)
-    local goodPlayer = PlayerResource:GetPlayer(goodGuy.id)
-    local badPlayer = PlayerResource:GetPlayer(badGuy.id)
-    local goodHero = goodPlayer:GetAssignedHero()
-    local badHero = badPlayer:GetAssignedHero()
-
-    goodGuy.duelNumber = 2
-    badGuy.duelNumber = 2
-
-    FindClearSpaceForUnit(goodHero, spawn1, true)
-    FindClearSpaceForUnit(badHero, spawn2, true)
-
-    Duels.zone2.addPlayer(goodGuy.id)
-    Duels.zone2.addPlayer(badGuy.id)
+  local playerSplitOffset = math.random(0, maxPlayers)
 
-    Duels:MoveCameraToPlayer(goodGuy.id, goodHero)
-    Duels:MoveCameraToPlayer(badGuy.id, badHero)
+  Duels:ShuffleTeams()
 
-    -- disable respawn
-    goodHero:SetRespawnsDisabled(true)
-    badHero:SetRespawnsDisabled(true)
+  local duelists = {{},{}}
+  for playerNumber = 1,playerSplitOffset do
+    for index, team in pairs(self.teams) do
+	    table.insert(duelists[index], team[playerNumber])
+	  end
   end
 
-  Duels.currentDuel = {
-    goodLiving1 = playerSplitOffset,
-    badLiving1 = playerSplitOffset,
-    goodLiving2 = maxPlayers - playerSplitOffset,
-    badLiving2 = maxPlayers - playerSplitOffset,
-    duelEnd1 = playerSplitOffset == 0,
-    duelEnd2 = maxPlayers == playerSplitOffset,
-    badPlayers = badPlayers,
-    goodPlayers = goodPlayers,
-    badPlayerIndex = badPlayerIndex,
-    goodPlayerIndex = goodPlayerIndex
-  }
-
-  Timers:CreateTimer(90, Dynamic_Wrap(Duels, 'EndDuel'))
-end
+  local spawns = Duels:GetSpawns(1)
+  self.duel1 = duelLib.Duel(duelists, spawns, self.zone1)
```

Comment:
Yes, lua does everything globally, no imports/exports or anything nice like that